serverlessworkflow/modules/ROOT/pages/eventing/working-with-callbacks.adoc

```diff
@@ -1 +1,108 @@
+
+[code,java]
+----
+
+    private String generateCloudEvent(String id, String input) {
+        Map<String, Object> eventBody = new HashMap<>();
+        eventBody.put("result", input + " and has been modified by the event publisher");
+        eventBody.put("dummyEventVariable", "This will be discarded by the process");
+        try {
+            return objectMapper.writeValueAsString(CloudEventBuilder.v1()
+                    .withId(UUID.randomUUID().toString())
+                    .withSource(URI.create(""))
+                    .withType("wait")
+                    .withTime(OffsetDateTime.now())
+                    .withExtension(CloudEventExtensionConstants.PROCESS_REFERENCE_ID, id)
+                    .withData(objectMapper.writeValueAsBytes(eventBody))
+                    .build());
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e);
+        }
+
+    }
+----
+
+The cloud event published by the listener modifies the original value of the `move` field and publish it as the `result` field. Moreover, it contains a header attribute, named `kogitoprocrefid`, holding the process instance id of the workflow that published the event. As mentioned before, this header is pivotal, because, when not using correlation, it is the only way for the callback state to know that this event should be used to resume the flow. 
+
+Remember that every workflow has a unique instance id that is automatically included in any event published by Kogito. As you can see in the code snippet below, the event listener takes the process instance id from a cloud event attribute named `kogitoprocinstanceid` of the event being consumed. 
+
+[source,java]
+----
+    @Incoming("in-resume")
+    @Outgoing("out-wait")
+    @Acknowledgment(Strategy.POST_PROCESSING)
+    public String onEvent(Message<String> message) {
+        Optional<CloudEvent> ce = CloudEventUtils.decode(message.getPayload());
+        JsonCloudEventData cloudEventData = (JsonCloudEventData) ce.get().getData();
+        return generateCloudEvent(ce.get().getExtension(CloudEventExtensionConstants.PROCESS_INSTANCE_ID).toString(), cloudEventData.getNode().get("move").asText());
+    }
+---- 
+
+=== Kafka configuration
+
+This example requires an external broker that handles the involved events in order to work. We are using link:{kafka_doc_url}[Kafka]. Kakfa use topics to publish/consume messages. In this example we are using two of them, which matches the names of event types defined in the workflow: `wait` and `resume`. They are configured in link:{kogito_sw_examples_url}/serverless-workflow-callback-quarkus/src/main/resources/application.properties}[application.properties]. If you are interested on details, take a look to link:{kogito_kafka_integration_url}[Kafka integration documentation]. 
```

Comment:
Maybe it's worth mentioning the Knative Eventing example too: https://github.com/kiegroup/kogito-docs/blob/main/serverlessworkflow/modules/ROOT/pages/eventing/consume-produce-events-with-knative-eventing.adoc

There, I'll link back to this guide that explains how to do callbacks in the Knative platform.

Reply:
It wont be better to link the knative example in further references? In this section we are just discussing how to setup Kakfa.

Reply:
Yeah, that's why I asked to add the Knative guide here as well. Because the Callback use case will be useful in both implementations. Just to not give the impression that we only support on Kafka.

Reply:
Ok, I will add that we have other alternative beyong Kafka here